File: Languages/Python/hashgen.py
print("Hello, World!")

import os
import hashlib
import json
import time
from pathlib import Path

def calculate_file_hash(file_path, algorithm='sha256'):
    """Calculate hash of a file using specified algorithm"""
    hash_obj = hashlib.new(algorithm)
    
    try:
        with open(file_path, 'rb') as f:
            # Read file in chunks to handle large files efficiently
            for chunk in iter(lambda: f.read(4096), b""):
                hash_obj.update(chunk)
        return hash_obj.hexdigest()
    except Exception as e:
        print(f"Error hashing {file_path}: {e}")
        return None

def hash_files_in_directory(directory_path):
    """Hash all files in the specified directory"""
    hashfiles_dir = directory_path
    output_file = "hash_results.json"
    
    if not os.path.exists(hashfiles_dir):
        print(f"Directory not found: {hashfiles_dir}")
        return
    
    print(f"Hashing all files in {hashfiles_dir}...")
    
    # Get all .txt files in the directory
    files = []
    for file_path in Path(hashfiles_dir).glob("*.txt"):
        files.append(file_path)
    
    files.sort()  # Sort for consistent ordering
    
    if not files:
        print("No .txt files found in directory")
        return
    
    print(f"Found {len(files)} files to hash")
    
    results = []
    successful_hashes = 0
    failed_hashes = 0
    
    start_time = time.time()
    
    for i, file_path in enumerate(files):
        file_hash = calculate_file_hash(file_path)
        
        result = {
            "filename": file_path.name,
            "filepath": str(file_path),
            "size": file_path.stat().st_size,
            "algorithm": "sha256"
        }
        
        if file_hash:
            result["hash"] = file_hash
            result["status"] = "SUCCESS"
            successful_hashes += 1
        else:
            result["hash"] = ""
            result["status"] = "FAILED"
            result["error"] = "Hash calculation failed"
            failed_hashes += 1
        
        results.append(result)
        
        # Progress indicator
        if (i + 1) % 100 == 0:
            print(f"Processed {i + 1}/{len(files)} files...")
    
    end_time = time.time()
    processing_time = end_time - start_time
    
    # Create final result
    final_result = {
        "total_files": len(files),
        "successful_hashes": successful_hashes,
        "failed_hashes": failed_hashes,
        "algorithm": "sha256",
        "processing_time_seconds": round(processing_time, 3),
        "average_time_per_file_ms": round((processing_time / len(files)) * 1000, 2),
        "timestamp": time.strftime("%Y-%m-%d %H:%M:%S"),
        "directory": hashfiles_dir,
        "results": results
    }
    
    # Save results to file
    try:
        with open(output_file, 'w') as f:
            json.dump(final_result, f, indent=2)
        print(f"Results saved to: {output_file}")
    except Exception as e:
        print(f"Error saving results: {e}")
    
    print(f"\nHashing completed!")
    print(f"Total files: {len(files)}")
    print(f"Successful: {successful_hashes}")
    print(f"Failed: {failed_hashes}")
    print(f"Processing time: {processing_time:.3f} seconds")
    print(f"Average time per file: {(processing_time / len(files) * 1000):.2f} ms")

if __name__ == "__main__":
    hash_files_in_directory("../hashfiles")

File: Languages/Rust/hashgen.rs
use std::collections::HashMap;
use std::fs;
use std::path::Path;
use std::time::Instant;
use sha2::{Sha256, Digest};
use serde::{Deserialize, Serialize};

// Hello, World!
fn hello() {
    println!("Hello, World!");
}

#[derive(Serialize, Deserialize, Debug)]
struct FileHashResult {
    filename: String,
    filepath: String,
    size: u64,
    hash: String,
    status: String,
    error: Option<String>,
}

#[derive(Serialize, Debug)]
struct HashResults {
    total_files: usize,
    successful_hashes: usize,
    failed_hashes: usize,
    algorithm: String,
    processing_time_seconds: f64,
    average_time_per_file_ms: f64,
    timestamp: String,
    directory: String,
    results: Vec<FileHashResult>,
}

fn calculate_file_hash(filepath: &str) -> Result<String, Box<dyn std::error::Error>> {
    let mut file = fs::File::open(filepath)?;
    let mut hasher = Sha256::new();
    let mut buffer = [0; 4096];
    
    loop {
        let bytes_read = file.read(&mut buffer)?;
        if bytes_read == 0 {
            break;
        }
        hasher.update(&buffer[..bytes_read]);
    }
    
    Ok(format!("{:x}", hasher.finalize()))
}

fn get_file_size(filepath: &str) -> u64 {
    fs::metadata(filepath).map(|m| m.len()).unwrap_or(0)
}

fn get_txt_files(directory: &str) -> Result<Vec<String>, Box<dyn std::error::Error>> {
    let mut files = Vec::new();
    
    for entry in fs::read_dir(directory)? {
        let entry = entry?;
        let path = entry.path();
        
        if path.is_file() && path.extension().map_or(false, |ext| ext == "txt") {
            files.push(path.to_string_lossy().to_string());
        }
    }
    
    files.sort();
    Ok(files)
}

fn get_timestamp() -> String {
    use chrono::Utc;
    Utc::now().format("%Y-%m-%d %H:%M:%S").to_string()
}

fn hash_files_in_directory(directory: &str) -> Result<(), Box<dyn std::error::Error>> {
    let files = get_txt_files(directory)?;
    
    if files.is_empty() {
        println!("No .txt files found in directory: {}", directory);
        return Ok(());
    }
    
    println!("Found {} files to hash", files.len());
    
    let mut results = HashResults {
        total_files: files.len(),
        successful_hashes: 0,
        failed_hashes: 0,
        algorithm: "sha256".to_string(),
        processing_time_seconds: 0.0,
        average_time_per_file_ms: 0.0,
        timestamp: get_timestamp(),
        directory: directory.to_string(),
        results: Vec::with_capacity(files.len()),
    };
    
    let start_time = Instant::now();
    
    for (i, filepath) in files.iter().enumerate() {
        let result = FileHashResult {
            filename: Path::new(filepath).file_name()
                .unwrap_or_default()
                .to_string_lossy()
                .to_string(),
            filepath: filepath.clone(),
            size: get_file_size(filepath),
            hash: String::new(),
            status: String::new(),
            error: None,
        };
        
        let mut final_result = result;
        
        match calculate_file_hash(filepath) {
            Ok(hash) => {
                final_result.hash = hash;
                final_result.status = "SUCCESS".to_string();
                results.successful_hashes += 1;
            }
            Err(_) => {
                final_result.hash = String::new();
                final_result.status = "FAILED".to_string();
                final_result.error = Some("Hash calculation failed".to_string());
                results.failed_hashes += 1;
            }
        }
        
        results.results.push(final_result);
        
        // Progress indicator
        if (i + 1) % 100 == 0 {
            println!("Processed {}/{} files...", i + 1, files.len());
        }
    }
    
    let end_time = start_time.elapsed();
    results.processing_time_seconds = end_time.as_secs_f64();
    results.average_time_per_file_ms = (results.processing_time_seconds / files.len() as f64) * 1000.0;
    
    // Save to file
    let output_data = serde_json::to_string_pretty(&results)?;
    fs::write("hash_results.json", output_data)?;
    
    println!("Results saved to: hash_results.json");
    println!("\nHashing completed!");
    println!("Total files: {}", results.total_files);
    println!("Successful: {}", results.successful_hashes);
    println!("Failed: {}", results.failed_hashes);
    println!("Processing time: {:.3} seconds", results.processing_time_seconds);
    println!("Average time per file: {:.2} ms", results.average_time_per_file_ms);
    
    Ok(())
}

fn main() {
    hello();
    
    if let Err(e) = hash_files_in_directory("../hashfiles") {
        eprintln!("Error: {}", e);
    }
}
